apps/designer/app/shared/stores/styles.ts

```diff
@@ -0,0 +1,108 @@
+import { useEffect } from "react";
+import store from "immerhin";
+import { createValueContainer, useValue } from "react-nano-state";
+import type { StyleProperty, StyleValue } from "@webstudio-is/css-data";
+import type { Styles } from "@webstudio-is/react-sdk";
+import { publish, subscribe } from "~/shared/pubsub";
+import { useSyncInitializeOnce } from "~/shared/hook-utils";
+
+export const stylesContainer = createValueContainer<Styles>([]);
+
+export const useStyles = () => useValue(stylesContainer);
+
+type SetStylesUpdate = {
+  store: "styles";
+  operation: "set";
+  breakpointId: string;
+  instanceId: string;
+  property: StyleProperty;
+  value: StyleValue;
+};
+
+type DeleteStylesUpdate = {
+  store: "styles";
+  operation: "delete";
+  breakpointId: string;
+  instanceId: string;
+  property: StyleProperty;
+};
+
+export type StylesUpdate = SetStylesUpdate | DeleteStylesUpdate;
+
+declare module "~/shared/pubsub" {
+  export interface PubsubMap {
+    loadStyles: Styles;
+  }
+}
+
+export const updateStyles = (styles: Styles, update: StylesUpdate) => {
+  const { breakpointId, instanceId, property } = update;
+  const matchedIndex = styles.findIndex(
+    (item) =>
+      item.breakpointId === breakpointId &&
+      item.instanceId === instanceId &&
+      item.property === property
+  );
+
+  if (update.operation === "set") {
+    const newItem = {
+      breakpointId,
+      instanceId,
+      property,
+      value: update.value,
+    };
+    if (matchedIndex === -1) {
+      styles.push(newItem);
+    } else {
+      styles[matchedIndex] = newItem;
+    }
+  }
+
+  if (update.operation === "delete" && matchedIndex !== -1) {
+    styles.splice(matchedIndex, 1);
+  }
+};
+
+export const useInitStyles = (styles: Styles) => {
+  // set initial styles value
+  useSyncInitializeOnce(() => {
+    stylesContainer.value = styles;
+  });
+
+  // styles are maintained on canvas side
+  // always send latest state to designer
+  useEffect(() => {
+    publish({
+      type: "loadStyles",
+      payload: stylesContainer.value,
+    });
+    stylesContainer.subscribe((styles) => {
+      publish({
+        type: "loadStyles",
+        payload: styles,
+      });
+    });
+  }, []);
+
+  // subscribe to styles updates
+  useEffect(() => {
+    return subscribe("update", (updates) => {
+      store.createTransaction([stylesContainer], (styles) => {
+        for (const update of updates) {
+          if (update.store === "styles") {
```

Comment:
lets talk about concept of stores you are now introducing. Initially nano-states were the same thing as stores, so what's the difference here.

I know we want to switch to nanostores, but this change is not doing it ....

Reply:
was it necessary to have an update specify the store `update.store` instead of just subscribing to "updateStyles" and setting the value on the nano state - this is how this has been done before, so I would like to understand what exactly we are changing and why

Reply:
update can include updates for single transaction. With events we'll get multiple transactions.

Reply:
> lets talk about concept of stores you are now introducing. Initially nano-states were the same thing as stores, so what's the difference here.

The idea is to group whole logic related to updating specific store/state. Caz it's getting messy with other nano states where all updates happen.

Once we rely only on specific messages we can add data indexing (some maps) for fast access.